src/network_service/sending_targets_cache.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::ConnectionInfo;
+use log::LogLevel;
+use std::collections::HashMap;
+use std::net::SocketAddr;
+
+const MAX_RESENDS: u8 = 3;
+
+enum TargetState {
+    Sending(u8),
+    Failed(u8),
+    Sent,
+}
+
+impl TargetState {
+    pub fn is_complete(&self) -> bool {
+        match *self {
+            TargetState::Failed(x) => x > MAX_RESENDS,
+            TargetState::Sent => true,
+            TargetState::Sending(_) => false,
+        }
+    }
+
+    pub fn is_sending(&self) -> bool {
+        match *self {
+            TargetState::Failed(_) | TargetState::Sent => false,
+            TargetState::Sending(_) => true,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct SendingTargetsCache {
+    cache: HashMap<u64, Vec<(ConnectionInfo, TargetState)>>,
+}
+
+impl SendingTargetsCache {
+    pub fn insert_message(
+        &mut self,
+        msg_id: u64,
+        initial_targets: Vec<ConnectionInfo>,
+        dg_size: usize,
+    ) {
+        let targets = initial_targets
+            .into_iter()
+            .enumerate()
+            .map(|(idx, tgt_info)| {
+                (
+                    tgt_info,
+                    if idx < dg_size {
+                        TargetState::Sending(0)
+                    } else {
+                        TargetState::Failed(0)
+                    },
+                )
+            })
+            .collect();
+        let _ = self.cache.insert(msg_id, targets);
+    }
+
+    fn target_states(&self, msg_id: u64) -> impl Iterator<Item = &(ConnectionInfo, TargetState)> {
+        self.cache.get(&msg_id).into_iter().flatten()
+    }
+
+    fn target_states_mut(
+        &mut self,
+        msg_id: u64,
+    ) -> impl Iterator<Item = &mut (ConnectionInfo, TargetState)> {
+        self.cache.get_mut(&msg_id).into_iter().flatten()
+    }
+
+    fn fail_target(&mut self, msg_id: u64, target: SocketAddr) {
+        let _ = self
+            .target_states_mut(msg_id)
+            .find(|(info, _state)| info.peer_addr() == target)
+            .map(|(_info, state)| match *state {
+                TargetState::Failed(_) => {
+                    log_or_panic!(LogLevel::Error, "Got a failure from a failed target!");
+                }
+                TargetState::Sending(x) => {
+                    *state = TargetState::Failed(x + 1);
```

Comment:
so this function will only increase the sending time by one, but won't update it's state to `Failed` ?
it's bit inconsistent with the function name, I think.

Reply:
It does update the state to `Failed`, though? I'm not sure what you mean.

Reply:
Ah, my fault, it does do the update.